### PR TITLE
New SEM!

### PR DESCRIPTION
**Pull request by vankesteren**
Finally, a new version of SEM! There are loads of improvements. 

New SEM has also fixed the following issues: [#440](https://github.com/jasp-stats/jasp-issues/issues/440), [#713](https://github.com/jasp-stats/jasp-issues/issues/713), [#834](https://github.com/jasp-stats/jasp-issues/issues/834), [#716](https://github.com/jasp-stats/jasp-issues/issues/716), [#671](https://github.com/jasp-stats/jasp-issues/issues/671), [#265](https://github.com/jasp-stats/jasp-issues/issues/265), [#735](https://github.com/jasp-stats/INTERNAL-jasp/issues/735), [#295](https://github.com/jasp-stats/INTERNAL-jasp/issues/295)

Touch-ups will be done next week, so don't merge yet.

### CODE REVIEW

File: tests/testthat/test-sem.R [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+  options <- jaspTools::analysisOptions("SEM")
+  options$models <- list(list(modelName = "Model1", syntax = "x1 ~ x2 + x3 + y1"))
+  options$emulation        <- "lavaan"
+  options$estimator        <- "default" 
+  options$groupingVariable <- ""
+  options$sampling.weights <- ""
+  options$information      <- "expected"
+  options$missing          <- "ML"
+  options$test             <- "standard"
+  results <- jaspTools::runAnalysis("SEM", "../poldem_grouped.csv", options)
```
Comment (TimKDJ) [inline tests/testthat/test-sem.R:13]: From what I gathered it's common to put the test resources in tests/testthat

File: R/common.R [hunk 2]
```diff
@@ -34,15 +34,87 @@ lavBootstrap <- function(fit, samples = 1000) {
     # and then every time on a successful bootstrap:
     # https://github.com/yrosseel/lavaan/blob/77a568a574e4113245e2f6aff1d7c3120a26dd90/R/lav_bootstrap.R#L375
     # i.e., samples + 1 times
-    progressbarTick()
+    JASP:::progressbarTick()
```
Comment (TimKDJ) [inline R/common.R:37]: Remove the namespace

File: R/common.R [hunk 3]
```diff
@@ -34,15 +34,87 @@ lavBootstrap <- function(fit, samples = 1000) {
...
     return(lavaan::coef(lav_object))
   }
-  startProgressbar(samples + 1)
+  JASP:::startProgressbar(samples + 1)
```
Comment (TimKDJ) [inline R/common.R:40]: And this one

File: NAMESPACE [hunk 4]
```diff
@@ -1,2 +1,4 @@
 export(SEM)
 export(MediationAnalysis)
+export(MIMIC)
```
Comment (TimKDJ) [inline NAMESPACE:3]: Can you add `import(jaspBase)` here?

File: DESCRIPTION [hunk 5]
```diff
@@ -1,8 +1,8 @@
 Package: SEM
 Type: Package
 Title: SEM Module for JASP
-Version: 0.0.0
-Date: 2020-04-22
+Version: 0.13.0
+Date: 2020-08-26
```
Comment (TimKDJ) [inline DESCRIPTION:5]: Could you add

```
Encoding: UTF-8
Imports:
  forcats,
  ggplot2,
  jaspBase,
  jaspGraphs,
  lavaan,
  reshape2,
  semPlot,
  semTools,
  stringr,
  tibble,
  tidyr
Remotes:
  jasp-stats/jaspBase,
  jasp-stats/jaspGraphs
```

File: R/common.R [hunk 6]
```diff
@@ -34,15 +34,87 @@ lavBootstrap <- function(fit, samples = 1000) {
...
+    ggplot2::scale_fill_continuous(low = "#FFFFFF", high = "#000000",
+                                   na.value = "transparent",
+                                   limits = c(0, 1)) +
+    ggplot2::coord_fixed() +
+    ggplot2::labs(x = "", y = "") +
+    ggplot2::theme(axis.ticks.x = ggplot2::element_blank()) +
+    ggplot2::theme(axis.ticks.y = ggplot2::element_blank()) +
+    ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90,
+                                                       hjust = 0)) +
+    JASPgraphs::themeJaspRaw()
```
Comment (TimKDJ) [inline R/common.R:77]: All `JASPgraphs` -> `jaspGraphs`

**Review by TimKDJ — CHANGES_REQUESTED** [review]

**Comment by vankesteren** [comment]
should be ready to merge now!

File: R/lgcm.R [hunk 7]
```diff
@@ -0,0 +1,793 @@
...
+  if (inherits(lgcmResult, "try-error")) {
+    modelContainer$setError(paste(
+      "Model error:", 
+      .decodeVarsInMessage(names(dataset), attr(lgcmResult, "condition")$message))
+    )
+    return()
+  }
+  
+  if (!lgcmResult@optim$converged) {
+    modelContainer$setError("The model could not be estimated due to nonconvergence.")
```
Comment (Kucharssim) [inline R/lgcm.R:111]: should this be inside `gettext()`?

File: R/lgcm.R [hunk 8]
```diff
@@ -0,0 +1,793 @@
...
+  }
+  
+  if (lgcmResult@test[[1]]$df < 0) {
+    modelContainer$setError("The model could not be estimated: No degrees of freedom left.")
```
Comment (Kucharssim) [inline R/lgcm.R:115]: `gettext()` again?

File: R/lgcm.R [hunk 9]
```diff
@@ -0,0 +1,793 @@
...
+    )
+  }
+  
+}
+
+.lgcmParameterTables <- function(modelContainer, dataset, options, ready) {
+  partabs <- if (!is.null(modelContainer[["partabs"]])) {
+    modelContainer[["partabs"]]
+  } else {
+    modelContainer[["partabs"]] <- createJaspContainer("Parameter estimates")
```
Comment (Kucharssim) [inline R/lgcm.R:255]: missing `gettext()`

Comment (Kucharssim) [reply R/lgcm.R:255]: these seem to be generally missing in this file

**Review by Kucharssim — COMMENTED** [review]
Looks neat! I left a couple of minor comments. I have trouble installing the module though on my `jasp-desktop/stable` and a nightly, so I was not able to check it out in action. Do I need to build `jasp-desktop/development` in order to run this module?

File: R/lgcm.R (hunk 9 above)
Comment (vankesteren) [reply R/lgcm.R:255]: oopsie

**Comment by vankesteren** [comment]
Using the latest nightly you should be able to install it, but because of the namespace change of jaspBase and JASP the analyses themselves will not work. [This branch](https://github.com/vankesteren/jaspSem/tree/ihnwhi) should work. It's the same (without the latest commit 6a0d36a stuff) but using the JASP namespace. It would be great if you could test it out!

**Review by Kucharssim — COMMENTED** [review]
Thanks @vankesteren! The branch worked - I was able to check it out, although I still could not see some plots because of `jaspGraphs`.

The module is looking fantastic though, the ability to compute and compare SEM models so easily is amazing, everything feels really smooth and nice.

I found two bugs/strange behaviour, see below.

#### First bug:
1. Open SEM and debug data set
2. write `X =~ contNormal + contGamma + contNarrow + contWide` in the model window and apply
3. The analysis stops with:
```
The following problem(s) occurred while running the analysis:
Infinity found in debInf
```
I believe the analysis should not check columns for errors that have not been specified in the model.

#### Second bug:
1. Open SEM and debug data set
2. Write `X =~ contNormal` in the model window and apply
3. The analysis stops with:
```
This analysis terminated unexpectedly.

Error in !ready: invalid argument type

Stack trace
tryCatchOne(expr, names, parentenv, handlers[[1]])

doTryCatch(return(expr), name, parentenv, handler)

withCallingHandlers(expr = analysis(jaspResults = jaspResults, dataset = dataset, options = options), error = .addStackTrace)

analysis(jaspResults = jaspResults, dataset = dataset, options = options)

SEM(...)

.semCheckErrors(dataset, options, ready, modelContainer)
```
The model is stupid of course, but it seems that the analysis did not catch this properly.

**Review by Kucharssim — APPROVED** [review]

**Comment by TimKDJ** [comment]
Hopefully this won't be your final contribution to JASP! ;)

**Pull request merged**